libs/langchain-google-common/src/chat_models.ts

```diff
@@ -60,14 +61,67 @@ class ChatConnection<AuthOptions> extends AbstractGoogleLLMConnection<
   BaseMessage[],
   AuthOptions
 > {
+
+  convertSystemMessageToHumanContent: boolean | undefined;
+
+  constructor(fields: GoogleAIBaseLLMInput<AuthOptions> | undefined, caller: AsyncCaller, client: GoogleAbstractedClient, streaming: boolean) {
+    super(fields, caller, client, streaming);
+    this.convertSystemMessageToHumanContent = fields?.convertSystemMessageToHumanContent;
+  }
+
+  get useSystemInstruction(): boolean {
+    return typeof this.convertSystemMessageToHumanContent === "boolean"
+      ? !this.convertSystemMessageToHumanContent
+      : this.computeUseSystemInstruction;
+  }
+
+  get computeUseSystemInstruction(): boolean {
+    // This works on models from April 2024 and later
+    //   Vertex AI: gemini-1.5-pro and gemini-1.0-002 and later
+    //   AI Studio: gemini-1.5-pro-latest
+    if (this.modelFamily === "palm") {
+      return false;
+    } else if (this.modelName === "gemini-1.0-pro-001") {
+      return false;
+    } else if (this.modelName.startsWith("gemini-pro-vision")) {
+      return false;
+    } else if (this.modelName.startsWith("gemini-1.0-pro-vision")) {
+      return false;
+    } else if (this.modelName === "gemini-pro" && this.platform === "gai") {
+      // on AI Studio gemini-pro is still pointing at gemini-1.0-pro-001
+      return false;
+    }
+    return true;
+  }
+
   formatContents(
     input: BaseMessage[],
     _parameters: GoogleAIModelParams
   ): GeminiContent[] {
     return input
-      .map((msg, i) => baseMessageToContent(msg, input[i - 1]))
-      .reduce((acc, cur) => [...acc, ...cur]);
+      .map((msg, i) =>
+        baseMessageToContent(msg, input[i - 1], this.useSystemInstruction)
+      )
+      .reduce((acc, cur) => {
+        // Filter out the system content, since those don't belong
+        // in the actual content.
+        const hasNoSystem = cur.every(content => content.role !== "system");
```

Comment:
Would we ever have system content mixed with other types of content?

Reply:
Gemini puts the system prompt in another section away from the content (but in a content-like data structure).

So `formatContents()` and `formatSystemInstruction()` both take the results of `baseMessageToContent()` and have to do different things with it.

In this function (`formatContents`) it removes all the system messages, since Gemini will reject the format if its there.
In the other function referenced above (`formatSystemInstruction`) it only needs to care if the system instruction is first.

Reply:
Technically we're dropping data silently in this case where a system message is not first in the input? I think logging a warning would be nice or ideally throwing an error.

Reply:
Latest update handles this by throwing an error. I've also added tests for this.